controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +124,296 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
-	log := ctrllog.FromContext(ctx)
-	log.Info("Running StorageClassClaim controller in Consumer Mode")
+func (r *StorageClassClaimReconciler) reconcileConsumerPhases() (reconcile.Result, error) {
+	r.log.Info("Running StorageClassClaim controller in Consumer Mode")
+
+	providerClient, err := providerclient.NewProviderClient(
+		r.ctx,
+		r.storageCluster.Spec.ExternalStorage.StorageProviderEndpoint,
+		10*time.Second,
+	)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Close client-side connections.
+	defer providerClient.Close()
+
+	if r.storageClassClaim.GetDeletionTimestamp().IsZero() {
+
+		// Check if finalizers are present, if not, add them.
+		if !contains(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer) {
+			storageClassClaimRef := klog.KRef(r.storageClassClaim.Name, r.storageClassClaim.Namespace)
+			r.log.Info("Finalizer not found for StorageClassClaim. Adding finalizer.", "StorageClassClaim", storageClassClaimRef)
+			r.storageClassClaim.SetFinalizers(append(r.storageClassClaim.GetFinalizers(), StorageClassClaimFinalizer))
+			if err := r.update(r.storageClassClaim); err != nil {
+				return reconcile.Result{}, fmt.Errorf("failed to update StorageClassClaim [%v] with finalizer: %s", storageClassClaimRef, err)
+			}
+		}
+
+		// Creation phase.
+		r.storageClassClaim.Status.Phase = ocsv1alpha1.StorageClassClaimCreating
+
+		// storageClassClaimStorageType is the storage type of the StorageClassClaim
+		var storageClassClaimStorageType providerclient.StorageType
+		switch r.storageClassClaim.Spec.Type {
+		case "blockpool":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestBlockpool
+		case "sharedfilesystem":
+			storageClassClaimStorageType = FulfillStorageClassClaimRequestSharedfilesystem
+		}
+
+		// Call the `FulfillStorageClassClaim` service on the provider server with StorageClassClaim as a request message.
+		_, err = providerClient.FulfillStorageClassClaim(
+			r.ctx,
+			r.storageCluster.Status.ExternalStorage.ConsumerID,
+			r.storageClassClaim.Name,
+			r.storageClassClaim.Spec.EncryptionMethod,
+			storageClassClaimStorageType,
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to initiate fulfillment of StorageClassClaim: %v", err)
+		}
+
+		// Configuration phase.
+		r.storageClassClaim.Status.Phase = ocsv1alpha1.StorageClassClaimConfiguring
+
+		// Call the `GetStorageClassClaimConfig` service on the provider server with StorageClassClaim as a request message.
+		response, err := providerClient.GetStorageClassClaimConfig(
+			r.ctx,
+			r.storageCluster.Status.ExternalStorage.ConsumerID,
+			r.storageClassClaim.Name,
+		)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("failed to get StorageClassClaim config: %v", err)
+		}
+		resources := response.ExternalResource
+		if resources == nil {
+			return reconcile.Result{}, fmt.Errorf("no configuration data recieved")
```

Comment:
use errors.New

Reply:
ACK

Reply:
Why? we are suing `fmt.Errorf` everywhere else. Why do we need to diverge here? just because the message is static? It does not matter, as even `fmt.Errorf` is using `errors.New` internally. 

And if we want to add a non-static part in the future then changing an `fmt.Errorf` is simpler than replacing `errors.New`, as `errors.New` will automatically raise the question of why was it used here? was there a reason specific reason? which will hider the delivery time for such a simple change.

Reply:
Fixed.